File: services/booking-bff/src/chaos.js
export const chaos = {
  error_rate: Number(process.env.CHAOS_ERROR_RATE ?? 0),
  latency_ms: Number(process.env.CHAOS_LATENCY_MS ?? 0),
};

export function registerChaos(app) {
  app.use(async (req, res, next) => {
    if (
      req.path.startsWith("/healthz") ||
      req.path.startsWith("/readyz") ||
      req.path.startsWith("/metrics") ||
      req.path.startsWith("/admin")
    ) {
      return next();
    }
    if (chaos.latency_ms) await new Promise((r) => setTimeout(r, chaos.latency_ms));
    if (chaos.error_rate && Math.random() < chaos.error_rate) {
      return res.status(500).json({ error: "injected failure" });
    }
    next();
  });

  app.post("/admin/chaos", (req, res) => {
    Object.assign(chaos, req.body);
    res.json(chaos);
  });

  app.get("/admin/chaos", (_req, res) => res.json(chaos));
}

File: services/inventory/app/chaos.py
import random
import threading
import time

from flask import jsonify, request


class ChaosState:
    def __init__(self):
        self.error_rate = 0.0
        self.latency_ms = 0
        self.db_pool_exhaust = False
        self.lock = threading.Lock()

    def snapshot(self):
        return {
            "error_rate": self.error_rate,
            "latency_ms": self.latency_ms,
            "db_pool_exhaust": self.db_pool_exhaust,
        }


chaos = ChaosState()


def register_chaos(app):
    @app.before_request
    def _apply_chaos():
        if request.path.startswith(("/healthz", "/readyz", "/metrics", "/admin")):
            return None
        if chaos.latency_ms:
            time.sleep(chaos.latency_ms / 1000.0)
        if chaos.error_rate and random.random() < chaos.error_rate:
            return jsonify({"error": "injected failure"}), 500
        return None

    @app.post("/admin/chaos")
    def _set_chaos():
        body = request.get_json(force=True)
        with chaos.lock:
            for k in ("error_rate", "latency_ms", "db_pool_exhaust"):
                if k in body:
                    setattr(chaos, k, body[k])
        return jsonify(chaos.snapshot())

    @app.get("/admin/chaos")
    def _get_chaos():
        return jsonify(chaos.snapshot())
